lib/appmap/trace.rb

```diff
@@ -37,6 +40,21 @@ def name
       def labels
         @package.labels
       end
+
+      private
+
+      # Read file and get last comment before line.
+      def self.last_comment(file, line)
+        File.open(file) do |f|
```

Comment:
The other implementation cached file contents.  Adding caching here might be only a few lines.

Reply:
`line` is used twice, as an input parameter and do parameter.

Reply:
The other implementation cached the whole file per method, which is a bit of a waste of memory. This one caches just the comment.

Reply:
I tried to test this with `method_source` testcases and got

```
$ ruby --version
ruby 3.0.1p64 (2021-04-05 revision 0fb782ee38) [x86_64-linux]

$bundle exec rspec spec

  1) MethodSource Methods should return a comment for method
     Failure/Error:
             f.lines.lazy.take(line_number - 1).reverse_each do |line|
               break unless (line =~ /^\s*#/) || (line =~ /^\s*$/)
       
               buffer << line.lstrip
             end
     
     NoMethodError:
       undefined method `lines' for #<File:/home/test/src/method_source/spec/spec_helper.rb (closed)>
       Did you mean?  lineno
     # ./lib/method_source.rb:52:in `block in comment_describing_custom'
     # ./lib/method_source.rb:49:in `open'
     # ./lib/method_source.rb:49:in `comment_describing_custom'
     # ./lib/method_source.rb:44:in `comment_helper'
     # ./lib/method_source.rb:143:in `comment'
     # ./spec/method_source_spec.rb:69:in `block (3 levels) in <top (required)>'
```

Maybe `f.lines` doesn't exist in Ruby 3.0?

Which version of Ruby did you use?

Reply:
In ruby 2.7.2 and ruby 3.0.1 it works with `f.each_line`.